pandas/src/inference.pyx

```diff
@@ -928,161 +923,160 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
                 break
         elif util.is_integer_object(val):
-            seen_int = 1
+            seen.int_ = 1
             floats[i] = <float64_t> val
             complexes[i] = <double complex> val
-            if not seen_null:
-                seen_uint = seen_uint or (int(val) > iINT64_MAX)
-                seen_sint = seen_sint or (val < 0)
+            if not seen.null_:
+                seen.saw_int(int(val))
 
-                if seen_uint and seen_sint:
-                    seen_object = 1
+                if seen.uint_ and seen.sint_:
+                    seen.object_ = 1
                     break
 
-                if seen_uint:
+                if seen.uint_:
                     uints[i] = val
-                elif seen_sint:
+                elif seen.sint_:
                     ints[i] = val
                 else:
                     uints[i] = val
                     ints[i] = val
 
         elif util.is_complex_object(val):
             complexes[i] = val
-            seen_complex = 1
+            seen.complex_ = 1
         elif PyDateTime_Check(val) or util.is_datetime64_object(val):
 
             # if we have an tz's attached then return the objects
             if convert_datetime:
                 if getattr(val, 'tzinfo', None) is not None:
-                    seen_datetimetz = 1
+                    seen.datetimetz_ = 1
                     break
                 else:
-                    seen_datetime = 1
+                    seen.datetime_ = 1
                     idatetimes[i] = convert_to_tsobject(
                         val, None, None, 0, 0).value
             else:
-                seen_object = 1
+                seen.object_ = 1
                 break
         elif try_float and not util.is_string_object(val):
             # this will convert Decimal objects
             try:
                 floats[i] = float(val)
                 complexes[i] = complex(val)
-                seen_float = 1
+                seen.float_ = 1
             except Exception:
-                seen_object = 1
+                seen.object_ = 1
                 break
         else:
-            seen_object = 1
+            seen.object_ = 1
             break
 
-    seen_numeric = seen_complex or seen_float or seen_int
+    seen.numeric_ = seen.complex_ or seen.float_ or seen.int_
 
     # we try to coerce datetime w/tz but must all have the same tz
-    if seen_datetimetz:
+    if seen.datetimetz_:
         if len(set([getattr(val, 'tzinfo', None) for val in objects])) == 1:
             from pandas import DatetimeIndex
             return DatetimeIndex(objects)
-        seen_object = 1
-
-    if not seen_object:
+        seen.object_ = 1
 
+    if not seen.object_:
         if not safe:
-            if seen_null:
-                if not seen_bool and not seen_datetime and not seen_timedelta:
-                    if seen_complex:
+            if seen.null_:
+                if (not seen.bool_ and not seen.datetime_
```

Comment:
it would be less code here if you would make this compound if as a property to Seen, though only if it has an intutie name, maybe ``is_float_or_complex`` (you might also need to pass ``safe`` to the Seen constructor